sycl/include/syclcompat/device.hpp

```diff
@@ -0,0 +1,553 @@
+        // by an int
+        get_info<sycl::info::device::max_work_item_sizes<3>>());
+#endif
+
+    prop.set_max_clock_frequency(
+        get_info<sycl::info::device::max_clock_frequency>());
+    prop.set_max_compute_units(
+        get_info<sycl::info::device::max_compute_units>());
+    prop.set_max_work_group_size(
+        get_info<sycl::info::device::max_work_group_size>());
+    prop.set_global_mem_size(get_info<sycl::info::device::global_mem_size>());
+    prop.set_local_mem_size(get_info<sycl::info::device::local_mem_size>());
+
+    size_t max_sub_group_size = 1;
+    std::vector<size_t> sub_group_sizes =
+        get_info<sycl::info::device::sub_group_sizes>();
+
+    for (const auto &sub_group_size : sub_group_sizes) {
+      if (max_sub_group_size < sub_group_size)
+        max_sub_group_size = sub_group_size;
+    }
+
+    prop.set_max_sub_group_size(max_sub_group_size);
+
+    prop.set_max_work_items_per_compute_unit(
+        get_info<sycl::info::device::max_work_group_size>());
+    int max_nd_range_size[] = {0x7FFFFFFF, 0x7FFFFFFF, 0x7FFFFFFF};
+    prop.set_max_nd_range_size(max_nd_range_size);
+
+    out = prop;
+  }
+
+  device_info get_device_info() const {
+    device_info prop;
+    get_device_info(prop);
+    return prop;
+  }
+
+  void reset() {
+    std::lock_guard<std::mutex> lock(m_mutex);
+    // The queues are shared_ptrs and the ref counts of the shared_ptrs increase
+    // only in wait_and_throw(). If there is no other thread calling
+    // wait_and_throw(), the queues will be destructed. The destructor waits for
+    // all commands executing on the queue to complete. It isn't possible to
+    // destroy a queue immediately. This is a synchronization point in SYCL.
+    _queues.clear();
+    // create new default queue.
+    _queues.push_back(
+        std::make_shared<sycl::queue>(_ctx, *this, detail::exception_handler,
+                                      sycl::property::queue::in_order()));
+    _saved_queue = _default_queue = _queues.front().get();
+  }
+
+  sycl::queue *default_queue() { return _default_queue; }
+
+  void queues_wait_and_throw() {
+    std::unique_lock<std::mutex> lock(m_mutex);
+    std::vector<std::shared_ptr<sycl::queue>> current_queues(_queues);
+    lock.unlock();
+    for (const auto &q : current_queues) {
+      q->wait_and_throw();
+    }
+    // Guard the destruct of current_queues to make sure the ref count is safe.
+    lock.lock();
+  }
+  queue_ptr create_queue(bool print_on_async_exceptions = false,
+                         bool in_order = true) {
+    std::lock_guard<std::mutex> lock(m_mutex);
+    sycl::async_handler eh = {};
+    sycl::property_list prop = {};
+    if (print_on_async_exceptions) {
+      eh = detail::exception_handler;
+    }
+    if (in_order) {
+      prop = {sycl::property::queue::in_order()};
+    }
+    _queues.push_back(std::make_shared<sycl::queue>(_ctx, *this, eh, prop));
```

Comment:
The way this is written if `print_on_async_exceptions` is false it will use the default initialized `async_handler`, which will replace the default behavior of calling `std::terminate` so async exceptions will fail silently, in this case do we want to initialize the `queue` without the async handler instead?

```
if (print_on_async_exceptions) {
  _queues.push_back(std::make_shared<sycl::queue>(_ctx, *this, detail::exception_handler, prop));
} else {
  _queues.push_back(std::make_shared<sycl::queue>(_ctx, *this, prop));
}
```